easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -677,43 +677,51 @@ def validate_iterate_opts_lists(self):
-
         faulty_deps = []
-        for i, hidden_dep in enumerate(self['hiddendependencies']):
+
+        for hidden_idx, hidden_dep in enumerate(self['hiddendependencies']):
             hidden_mod_name = ActiveMNS().det_full_module_name(hidden_dep)
             visible_mod_name = ActiveMNS().det_full_module_name(hidden_dep, force_visible=True)
 
-            # track whether this hidden dep is listed as a build dep
-            if visible_mod_name in build_dep_mod_names or hidden_mod_name in build_dep_mod_names:
-                # templating must be temporarily disabled when updating a value in a dict;
-                # see comments in resolve_template
-                enable_templating = self.enable_templating
-                self.enable_templating = False
-                self['hiddendependencies'][i]['build_only'] = True
-                self.enable_templating = enable_templating
-
-            # filter hidden dep from list of (build)dependencies
-            if visible_mod_name in dep_mod_names:
-                for key in ['builddependencies', 'dependencies']:
-                    self[key] = [d for d in self[key] if d['full_mod_name'] != visible_mod_name]
-                self.log.debug("Removed (build)dependency matching hidden dependency %s", hidden_dep)
-            elif hidden_mod_name in dep_mod_names:
-                for key in ['builddependencies', 'dependencies']:
-                    self[key] = [d for d in self[key] if d['full_mod_name'] != hidden_mod_name]
-                self.log.debug("Hidden (build)dependency %s is already marked to be installed as a hidden module",
-                               hidden_dep)
-            else:
+            # replace (build) dependencies with their equivalent hidden (build) dependency (if any)
+            replaced = False
+            for key in ['builddependencies', 'dependencies']:
+                for idx, dep in enumerate(self[key]):
+                    dep_mod_name = dep['full_mod_name']
+                    if dep_mod_name in [visible_mod_name, hidden_mod_name]:
+
+                        # templating must be temporarily disabled to obtain reference to original lists,
+                        # so their elements can be changed in place
+                        # see comments in resolve_template
+                        enable_templating = self.enable_templating
+                        self.enable_templating = False
+
+                        # track whether this hidden dep is listed as a build dep
+                        orig_hidden_dep = self['hiddendependencies'][hidden_idx]
+                        if key == 'builddependencies':
+                            orig_hidden_dep['build_only'] = True
+                        # actual replacement
+                        self[key][idx] = orig_hidden_dep
+                        replaced = True
+
+                        self.enable_templating = enable_templating
```

Comment:
@bartoldeman Style nitpicking: do this in the block above, after the `self[key][idx]`, since that's where it actually matters?

So, drop empty line above, add one below this line, to make it a bit less confusing.